aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsEndpointDecorator.kt

```diff
@@ -123,3 +138,202 @@ class PubUseEndpoint(private val runtimeConfig: RuntimeConfig) : LibRsCustomizat
         }
     }
 }
+
+class EndpointResolverGenerator(protocolConfig: ProtocolConfig, private val endpointData: ObjectNode) {
+    private val runtimeConfig = protocolConfig.runtimeConfig
+    private val endpointPrefix = protocolConfig.serviceShape.expectTrait<ServiceTrait>().endpointPrefix
+    private val awsEndpoint = runtimeConfig.awsEndpointDependency().asType()
+    private val codegenScope =
+        arrayOf(
+            "Partition" to awsEndpoint.member("Partition"),
+            "endpoint" to awsEndpoint.member("partition::endpoint"),
+            "CredentialScope" to awsEndpoint.member("CredentialScope"),
+            "Regionalized" to awsEndpoint.member("partition::Regionalized"),
+            "Protocol" to awsEndpoint.member("partition::endpoint::Protocol"),
+            "SignatureVersion" to awsEndpoint.member("partition::endpoint::SignatureVersion")
+        )
+
+    fun resolver(): RuntimeType {
+        val partitionsData = endpointData.expectArrayMember("partitions").getElementsAs(Node::expectObjectNode)
+
+        val comparePartitions = object : Comparator<PartitionNode> {
+            override fun compare(x: PartitionNode, y: PartitionNode): Int {
+                // always sort standard aws partition first
+                if (x.id == "aws") return -1
+                return x.id.compareTo(y.id)
+            }
+        }
+
+        val partitions = partitionsData.map {
+            PartitionNode(endpointPrefix, it)
+        }.sortedWith(comparePartitions)
+        val base = partitions.first()
+        val rest = partitions.drop(1)
+        val fnName = "endpoint_resolver"
+        return RuntimeType.forInlineFun(fnName, "aws_endpoint") {
+            it.rustBlock("pub fn $fnName() -> impl #T", awsEndpoint.member("ResolveAwsEndpoint")) {
```

Comment:
Nit: I think these will be more readable if you just use rustBlockTemplate, name the types, and pass in `*codegenScope`.